lisa/wlgen/workload.py

```diff
@@ -72,24 +72,18 @@ def run(self, cpus=None, cgroup=None, background=False, as_root=False, value=42)
     required_tools = ['taskset']
     """
     The tools required to execute the workload. See
-    :meth:`lisa.env.TestEnv.install_tools`.
+    :meth:`lisa.target.TestEnv.install_tools`.
     """
 
-    def __init__(self, te, name, res_dir=None):
-        self.te = te
+    def __init__(self, target, name, res_dir):
+        self.target = target
         self.name = name
-
-        # XXX: move this to run() instead?
-        if not res_dir:
```

Comment:
Why that change? I've tried running a simple notebook and I end being forced to specify a directory, I suppose this stems from `get_res_dir()` being removed? It was a very convenient feature.

Reply:
The target is not a res dir manager anymore, that's a target. That responsibility is delegated to ArtifactPathManager, which has a `get_res_dir()` method for interactive use. If needed, we can have an alternate constructor that would take a TestEnv-ish object and use it like we used to do. The old way of passing a TestEnv to `__init__` is akin to having `__init__` taking a configuration object instead of a bunch of parameters: the basic API takes a bunch of parameters, and some alternate constructors provide sugar on top of that if needed.

Reply:
> 
> ```
>         res_dir = res_dir or ArtifactPathManager().get_dir(
>             name=cls.__qualname__,
>             symlink=True,
>         )
> ```

Looks fine to me - I don't care who provides the directory as long as I don't have to care about it if I don't care about it (heh).

Reply:
If we end up having more things like that, it could be a good idea to consolidate all that interactive-style use in a `TestEnv`, and having a `from_testenv` method extracting all the required bits out of it.
The critical part it: that `TestEnv` role is only to provide better interactive support, so it can only be manipulated by toplevel methods dedicated to that. Lower-level APIs must use independent pieces.